src/optimizer/rule_impls.cpp

```diff
@@ -111,18 +128,18 @@ void GetToSeqScan::Transform(
 ///////////////////////////////////////////////////////////////////////////////
 /// GetToIndexScan
 GetToIndexScan::GetToIndexScan() {
-  physical = true;
+  type_ = RuleType::GET_TO_INDEX_SCAN;
 
   match_pattern = std::make_shared<Pattern>(OpType::Get);
 }
 
 bool GetToIndexScan::Check(std::shared_ptr<OperatorExpression> plan,
-                           Memo *memo) const {
+                           OptimizeContext *context) const {
   // If there is a index for the table, return true,
   // else return false
-  (void)memo;
-  bool index_exist = false;
+  (void)context;
   const LogicalGet *get = plan->Op().As<LogicalGet>();
+  bool index_exist = false;
   if (get != nullptr && get->table != nullptr &&
       !get->table->GetIndexColumns().empty())
```

Comment:
One minor question. The [`CatalogCache`](https://github.com/cmu-db/peloton/blob/master/src/include/catalog/catalog_cache.h#L27) object's members are all private, is it ok to directly change those getter/setter to public? 
In fact, I didn't find other parts of the system using this class or setting up this member in the `TransactionContext`. Do we also want to work on initializing this object in our system in this PR?

Reply:
You can set the getters to public but not the setters. The setters are designed to be private.

The Binder and the Optimizer are exactly the places that are going to use this class. It is already set up [here](https://github.com/cmu-db/peloton/blob/master/src/include/concurrency/transaction_context.h#L149). You don't need to initialize anything to use it. All the members are designed to be lazily initialized the first time referenced. You can just use it as if all the information is already there.

Reply:
I find two things when trying to use the `CatalogCache` in the `TransactionContext`. 

Firstly, I think we might be initializing the `CatalogCache` using the [default constructor](https://github.com/cmu-db/peloton/blob/master/src/include/catalog/catalog_cache.h#L37) in the `TransactionContext` so there's nothing in it. When I tried to find the `DatabaseCatalogObject`, I always get a nullptr.

Another thing is that we may also want to provide an interface to get the cached function table since we bind the function in the binder [`here`](https://github.com/cmu-db/peloton/blob/master/src/binder/bind_node_visitor.cpp#L209).

Reply:
I see. I think both of the two questions are for @mengranwo.

Mengran - 
1. The [`catalog_cache`](https://github.com/cmu-db/peloton/blob/master/src/include/concurrency/transaction_context.h#L149) is indeed uninitialized in the transaction context. How are they supposed to access the information of the catalog objects?
2. How should they access the function table, as Bowei said?

Maybe you can provide some quick fix for them if those are not too difficult to fix?